### PR TITLE
refactor: replace PersistState with IngestState

### PR DESCRIPTION
**Pull request by domodwyer**
Decouples the way the persist system says "don't accept writes" and generalises it for later use for gracefully stopping ingest.

Pre-requisite for #6549, which will make use of the new error state `GracefulShutdown`. This is configured to return a `PRECONDITION_FAILED` response (message: `ingester is shutting down"`) for write requests when set.

---

* refactor: replace PersistState with IngestState (63b1e4a3a)

      Prior to this commit, when initialising the persist system it would
      return a PersistState instance, used to communicate the saturation
      status of the persistence system. The RPC write path used this
      information to accept or deny write requests accordingly.
      
      This was unfortunate in that it tightly coupled the ingest handler to
      the persist system - in order to initialise the RPC handler, you had to
      provide a PersistState; this required us to initialise a persist system
      when testing only the RPC handler (which had nothing to do with
      persisting). This smells!
      
      This commit inverts the dependency, and decouples the subsystems via a
      shared type (IngestState). Instead of the persist system telling ingest
      to stop, the ingest system provides a means to be told to stop - this
      subtle difference decouples the ingest handler from all components that
      need to block ingest. This allows a fast O(1) error state read for N
      components and prevents us from having to start N components to test a
      RPC handler.
      
      Additionally this commit introduces an unused ingest error state
      (GracefulStop) as part of figuring out the API (to be used shortly).

### CODE REVIEW

File: ingester2/src/ingest_state.rs [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+/// subsystems to the ingest code path. Many error states can be set at any one
+/// time, but at most one error is returned by a call to
+/// [`IngestState::read()`].
+///
+/// Calling [`IngestState::read()`] returns [`Ok`], or the error currently set
+/// with [`IngestState::set()`].
+///
+/// Error states can be unset by calling [`IngestState::unset()`].
+#[derive(Debug, Default)]
+pub(crate) struct IngestState {
```
Comment (domodwyer) [inline ingester2/src/ingest_state.rs:40]: This is the type that allows subsystems to set the system state.

You call `IngestState::set()` with an error, and it is returned by calls to [`IngestState::read()`]. The error state can be unset at any point (which obviously won't happen for graceful stops).

File: ingester2/src/ingest_state.rs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    ///
+    /// # Precedence
+    ///
+    /// If more than one error state it set, a single error is returned based on
+    /// the following precedence (ordered by highest priority to lowest):
+    ///
+    ///   1. [`IngestStateError::GracefulStop`]
+    ///   2. [`IngestStateError::PersistSaturated`].
+    ///
+    pub(crate) fn read(&self) -> Result<(), IngestStateError> {
```
Comment (domodwyer) [inline ingester2/src/ingest_state.rs:132]: This describes how N errors are mapped down to 1 read error.

An RPC request can only return 1 error, so this documents the precedence handling.

File: ingester2/src/init.rs [hunk 3]
```diff
@@ -217,11 +218,17 @@ pub async fn new(
...
+    // write path.
+    let ingest_state = Arc::new(IngestState::default());
+
     // Spawn the persist workers to compact partition data, convert it into
     // Parquet files, and upload them to object storage.
-    let (persist_handle, persist_state) = PersistHandle::new(
+    let persist_handle = PersistHandle::new(
         persist_workers,
         persist_queue_depth,
+        Arc::clone(&ingest_state),
```
Comment (domodwyer) [inline ingester2/src/init.rs:231]: The `IngestState` is _passed into_ the persist system...

File: ingester2/src/init.rs [hunk 4]
```diff
@@ -286,7 +293,7 @@ pub async fn new(
             Arc::new(write_path),
             buffer,
             timestamp,
-            persist_state,
+            ingest_state,
```
Comment (domodwyer) [inline ingester2/src/init.rs:296]: ... and _passed into_ the RPC handlers.

This ensures they're fully decoupled and prevents horrible false testing dependencies.

File: ingester2/src/server/grpc/rpc_write.rs [hunk 5]
```diff
@@ -251,11 +247,9 @@ mod tests {
                     );
                     let timestamp = Arc::new(TimestampOracle::new(0));
 
-                    let metrics = metric::Registry::default();
-                    let sem = Arc::new(Semaphore::new(PERSIST_QUEUE_DEPTH));
-                    let persist_state = Arc::new(PersistState::new(PERSIST_QUEUE_DEPTH, sem, &metrics));
+                    let ingest_state = Arc::new(IngestState::default());
 
-                    let handler = RpcWrite::new(Arc::clone(&mock), timestamp, persist_state);
```
Comment (domodwyer) [inline ingester2/src/server/grpc/rpc_write.rs:258]: We no longer have to start the persist system, provide it with a metrics instance, and initialise a semaphore (none of which are used) just to test the RPC handler 🎉

File: ingester2/src/server/grpc/rpc_write.rs [hunk 6]
```diff
@@ -463,7 +453,7 @@ mod tests {
             .await
             .expect("write should succeed");
 
-        persist_state.test_set_state(CurrentState::Saturated);
```
Comment (domodwyer) [inline ingester2/src/server/grpc/rpc_write.rs:466]: The decoupling means there's no need to use a horrible test-only hack were we reach in and falsely set the state of the persist system (which breaks invariants of the persist system!) just to drive the RPC handler logic.

File: ingester2/src/server/grpc/rpc_write.rs [hunk 7]
```diff
@@ -38,21 +38,18 @@ enum RpcError {
...
+    SystemState(IngestStateError),
 }
 
 impl From<RpcError> for tonic::Status {
     fn from(e: RpcError) -> Self {
         let code = match e {
             RpcError::Decode(_) | RpcError::NoPayload | RpcError::NoTables => Code::InvalidArgument,
-            RpcError::PersistSaturated => Code::ResourceExhausted,
+            RpcError::SystemState(IngestStateError::PersistSaturated) => Code::ResourceExhausted,
+            RpcError::SystemState(IngestStateError::GracefulStop) => Code::FailedPrecondition,
```
Comment (domodwyer) [inline ingester2/src/server/grpc/rpc_write.rs:52]: This is the mapping of ingest state errors to RPC responses.

File: ingester2/src/server/grpc/rpc_write.rs [hunk 8]
```diff
@@ -476,4 +466,60 @@ mod tests {
         // One write should have been passed through to the DML sinks.
         assert_matches!(*mock.get_calls(), [DmlOperation::Write(_)]);
     }
+
+    /// Validate that the ingester being marked as stopping prevents the
+    /// ingester from accepting new writes.
+    #[tokio::test]
+    async fn test_rpc_write_shutdown() {
```
Comment (domodwyer) [inline ingester2/src/server/grpc/rpc_write.rs:473]: This tests the RPC handler when the system is marked as gracefully stopping.

**Review by pauldix — APPROVED** [review]

**Pull request merged**